dockerfile-template/Dockerfile

```diff
@@ -1,32 +1,60 @@
-# base image
-FROM ubuntu:focal
+# FROM defines the base docker image. This command has to come first in the file
+# The 'as' keyword lets you name the folowing stage. We use `app` for the production image
+FROM ubuntu:focal as app
 
-# metadata - optional, but highly recommended
+# ARG sets environment variables during the build stage
+ARG SOFTWARENAME_VER="1.0.0"
+
+# LABEL instructions tag the image with metadata that might be important to the user
+# Optional, but highly recommended
 LABEL base.image="ubuntu:focal"
 LABEL dockerfile.version="1"
 LABEL software="SoftwareName"
-LABEL software.version="1.0.0"
+LABEL software.version=$SOFTWARENAME_VER
 LABEL description="This software does X, Y, AND Z!"
 LABEL website="https://github.com/StaPH-B/docker-builds"
 LABEL license="https://github.com/StaPH-B/docker-builds/blob/master/LICENSE"
 LABEL maintainer="FirstName LastName"
 LABEL maintainer.email="my.email@email.com"
 
-# install dependencies via apt-get or yum if using a centos or fedora base 
+# RUN executes code during the build
+# Install dependencies via apt-get or yum if using a centos or fedora base
 RUN apt-get update && apt-get install -y \
- wget \
- perl
+ dependency_a \
+ dependency_b \
+ dependency_c
 
-# install and/or setup more things. Make /data for use as a working dir
+# Install and/or setup more things. Make /data for use as a working dir
 # Example: ncbi-blast+ 2.9.0
 RUN wget ftp://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/2.9.0/ncbi-blast-2.9.0+-x64-linux.tar.gz && \
  tar -xzf ncbi-blast-2.9.0+-x64-linux.tar.gz && \
  rm ncbi-blast-2.9.0+-x64-linux.tar.gz && \
  mkdir /data
 
-# set environmental variables e.g. $PATH and locale settings for singularity compatibility
+# ENV instructions set environment variables that persist from the build into the resulting image
+# Use for e.g. $PATH and locale settings for compatibility with Singularity
 ENV PATH="/ncbi-blast-2.9.0+/bin:$PATH" \
  LC_ALL=C
 
-# set working directory
+# WORKDIR sets working directory
 WORKDIR /data
+
+# A second FROM insruction creates a new stage
+# We use `test` for the test image
+FROM app as test
+
+# Demonstrate that the program is successfully installed
+
+# Option 1: run the program's internal tests, for example with SPAdes:
+RUN spades.py --test
+
+# Option 2: write your own tests in a bash script in the same directory as your Dockerfile:
+COPY my_tests.sh .
+RUN bash my_tests.sh
```

Comment:
I'm a little worried about adding custom files to these dockerfiles. The way docker images build requires a certain amount of space for the end user. 

For example, spades is meant to create a de novo consensus sequence from fastq files, which are large files.

Reply:
If I get your concern right, multistage dockerfiles help. Any data and layers added in the "test" stage are isolated from the "app" stage. So, building to the test stage is only done on the github actions runners, the test stage image is never pushed to Docker Hub. Building to the app stage yields a nice small image to push to Docker Hub.

Reply:
That is so fancy 😵